### ISSUE TITLE
connlib: Replace `webrtc-rs`

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#1962]
`webrtc-rs` suffers from a slew of issues that are difficult to address or work around. Further, the library is much, much larger than we need. Instead, @thomaseizinger has recommended looking at rebuilding our NAT traversal implementation using the better-architected, lower-level [str0m](https://github.com/algesten/str0m) library.

Because this functionality is core to our product, it makes sense to invest the time to do it well. This issue exists to track work related to:

- **Do we even need the whole WebRTC standard at all**? If not, str0m may not be a good fit.
- If so, itemize the work required to replace `webrtc-rs` with `str0m`

**Comment by thomaseizinger** [comment]
> * **Do we even need the whole WebRTC standard at all**?

I think it makes sense to pick this apart a bit. WebRTC is a collection of standards with ICE f.e. being separate from TURN being separate from video streams etc.

It is worth considering building on top of ICE to establishing p2p wireguard connections or relay them via TURN on hole-punch failures. I don't think this needs to happen via WebRTC data channels and neither do we need any media streams etc.

A quick google search turned up https://github.com/cunicu/cunicu which we could take as inspiration. What we should consider here is that we will be sending plain wireguard traffic which can easily be identified I think and might be blocked by certain actors. Those might exactly be the customers that rely on a VPN. I don't have a solution for that but I think we can consider that to be a separate issue because unless we double-encrypt, the traffic will always be wire-guard traffic. Double encryption hurts performance so we should probably avoid it when we can.

> If so, itemize the work required to replace `webrtc-rs` with `str0m`

I would need to look into it in more detail but we may be able to reuse the ICE state machine of `str0m` to do the hole punching for us and then send the wireguard traffic over it. If not, we can still fork it / write our own.

**Comment by thomaseizinger** [comment]
The following article also discusses something similar: https://nordvpn.com/blog/achieving-nat-traversal-with-wireguard/

**Comment by jamilbk** [comment]
> What we should consider here is that we will be sending plain wireguard traffic which can easily be identified I think and might be blocked by certain actors.

Yes, this is actually a business requirement for us: some customers have employees that occasionally need to travel to countries that outright block WireGuard using [DPI systems](https://www.ntkernel.com/how-to-bypass-egypts-wireguard-ban/) to identify the handshake.

The good news is that we can circumvent the block. We've discussed this internally, and we can probably get away with TURN-TCP for v1. If that fails, then we can wrap it with TLS instead.

We brainstormed a more advanced algorithm [in this issue](https://github.com/firezone/product/issues/528), and would love to dig in further. But for now, simple TCP wrapping would probably suffice for a first iteration.

> I would need to look into it in more detail but we may be able to reuse the ICE state machine of str0m to do the hole punching for us and then send the wireguard traffic over it. If not, we can still fork it / write our own.

👍

**Comment by jamilbk** [comment]
@thomaseizinger One advantage to rolling our own holepunching solution from scratch is that we can take advantage of certain optimizations to improve the connectivity experience.

- Perhaps we can use the first packet of the WireGuard handshake as the probe, then essentially use a variant of ICE Trickle to automatically pick the first channel that completes the handshake successfully.
- We can customize the ICE gathering process to inject candidates we know will have a higher chance of succeeding. For example, we could designate a port range users can open to their Gateways that will be tried first.

**Comment by thomaseizinger** [comment]
> We can customize the ICE gathering process to inject candidates we know will have a higher chance of succeeding. For example, we could designate a port range users can open to their Gateways that will be tried first.

This is "supported" by `str0m` because it is SANS-IO: https://github.com/algesten/str0m?tab=readme-ov-file#nic-enumeration-and-turn-and-stun

**Comment by thomaseizinger** [comment]
One of our users over at `rust-libp2p` apparently has a working implementation of the libp2p `webrtc-direct` protocol using `str0m`. See https://github.com/algesten/str0m/issues/166#issuecomment-1703346556.

I don't know when / if they'll open a PR but it should give us a good template for how we could integrate `str0m` into firezone.

**Comment by thomaseizinger** [comment]
> **Do we even need the whole WebRTC standard at all**? If not, str0m may not be a good fit.

So, it depends a bit on how far you want to take this but here is an idea:

ICE by itself is protocol-agnostic, it just specifies the state machine for establishing p2p connections including NAT traversal. We could take a module like https://github.com/algesten/str0m/tree/main/src/ice, feed it local candidates and remote candidates (i.e. the TURN server). The portal would be used as a signalling server to exchange the candidates.

We then have the state machine (i.e. the `ice` module of `str0m`) guide use through the signalling checks of the various candidates. This should either result in a hole-punch or a fallback to the TURN server plus an allocation / channel.

Once that is established, we should be able to "just" send UDP packets to a certain peer via the address that the ICE agent yielded (this could be a relay address allocated by TURN).

For me, there are two open questions:

1. Somehow, we need to keep the ICE agent informed about the state of the connection. In the SANS-IO implementation of `str0m`, this probably means to keep calling it as part of some eventloop that also transfers the data / interacts with the connection.
2. I am not familiar with how our wireguard libray works but what I think we'd need from it is that it also operates in a SANS-IO fashion where it only yields the packets that we should send so that we can own the actual socket. That is because we'd keep the socket to a particular gateway separately. In other words, imagine an event-loop that owns three components: an ICE agent. a wireguard endpoint, a UDP socket. We'd establish a connection to the gateway but consulting the ICE agent and allocate a socket accordingly to the result. Once established, the wireguard endpoint would be "live" and yield encrypted packets that we have to send, and all we'd do is forward them to the UDP socket.

I think what is currently not handled by `str0m` (see https://github.com/algesten/str0m/issues/325) is that TURN state machine of creating allocations and refreshing them incl. the permissions / data channels. That would also have to be part of the above described event loop.

What could be useful is to develop a library that implements ICE + STUN + TURN in a SANS-IO fashion, i.e. performs connectivity checks based on given candidates. If those fail, creates[^1] a TURN allocation and yields you the address as a connection. Candidate exchange would also be SANS-IO such that what you use as the signaling server (in our case the portal) is up to you.

What you send over said connection is completely up-to-you, which means it is completely ambivalent of wireguard's existence.

[^1]: In SANS-IO speak, this means it tells you do send a certain packet to a TURN server)

**Comment by thomaseizinger** [comment]
One of our community members has created an oppinionated fork of `rust-libp2p` and is using `str0m` for the webrtc transport. See https://github.com/altonen/litep2p/blob/master/src/transport/webrtc.

Could be helpful as inspiration :)

**Comment by thomaseizinger** [comment]
Closing in favor of https://github.com/firezone/firezone/issues/2966.
